src/rust/aggregator/src/pipelines/order_history.rs

```diff
@@ -0,0 +1,324 @@
+        )
+        .fetch_all(transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?;
+
+        // Insert all places as orders into the state.
+        for place in places {
+            orders.insert((place.market_id.clone(), place.order_id.clone()), place);
+        }
+
+        // Remove all orders that have been cancelled.
+        for cancel in cancels {
+            orders.remove(&(cancel.market_id, cancel.order_id));
+        }
+
+        // Handle fills and changes chronologically.
+        let mut fills_index = 0;
+        let mut changes_index = 0;
+        let handle_fill =
+            |fill: &Fill, orders: &mut HashMap<OrderKey, Order>, index: &mut usize| {
+                *index += 1;
+                let order_key_maker = (fill.market_id.clone(), fill.maker_order_id.clone());
+                if let Some(order) = orders.get_mut(&order_key_maker) {
+                    order.size -= fill.size.clone();
+                    if order.size.is_zero() {
+                        orders.remove(&order_key_maker);
+                    }
+                }
+                let order_key_taker = (fill.market_id.clone(), fill.taker_order_id.clone());
+                if let Some(order) = orders.get_mut(&order_key_taker) {
+                    order.size -= fill.size.clone();
+                    if order.size.is_zero() {
+                        orders.remove(&order_key_taker);
+                    }
+                }
+            };
+        let handle_change =
+            |change: &Change, orders: &mut HashMap<OrderKey, Order>, index: &mut usize| {
+                *index += 1;
+                let order_key = (change.market_id.clone(), change.order_id.clone());
+                if let Some(order) = orders.get_mut(&order_key) {
+                    order.size = change.new_size.clone();
+                    if order.size.is_zero() {
+                        orders.remove(&order_key);
+                    }
+                }
+            };
+        while fills_index < fills.len() && changes_index < changes.len() {
+            let fill = fills.get(fills_index).unwrap();
+            let fill_index = (fill.txn_version.clone(), fill.event_idx.clone());
+
+            let change = changes.get(changes_index).unwrap();
+            let change_index = (change.txn_version.clone(), change.event_idx.clone());
+
+            if fill_index < change_index {
+                handle_fill(fill, &mut orders, &mut fills_index);
+            } else {
+                handle_change(change, &mut orders, &mut changes_index);
+            }
+        }
+        while fills_index < fills.len() {
+            let fill = fills.get(fills_index).unwrap();
+            handle_fill(fill, &mut orders, &mut fills_index);
+        }
+        while changes_index < changes.len() {
+            let change = changes.get(changes_index).unwrap();
+            handle_change(change, &mut orders, &mut changes_index);
+        }
```

Comment:
I think this loop could be more effectively abstracted to only have one call each to `handle_fill` and `handle_change`

How about making it more like the `user_history` call that starts

```rust
for _ in 0..(fill_events.len() + change_events.len())
```

Reply:
The new logic is more abstracted, but it still has two calls each to `handle_fill` and `handle_change` that could be consolidated for easier maintainability

How about using the `let (fill_event_to_aggregate, change_event_to_aggregate)` paradigm as in the original algorithm, ending with the `mach(fill_event_to_aggregator, change_event_to_aggregate)` statement that would, in this case, result in a *single* call to `handle_fill` and `handle_change`?

Reply:
I'm sorry but I do not get what you mean. Furthermore, since `handle_*` are functions, I do not see how calling them twice would affect maintainability.

Reply:
Noting offline discussion to incorporate the `_to_aggregate` option paradigm for consolidation of duplicate function calls